src/PersistorHelper.js

```diff
@@ -0,0 +1,186 @@
+const Crypto = require('crypto')
+const Stream = require('stream')
+const Settings = require('./Settings')
+const Logger = require('logger-sharelatex')
+const { WriteError, ReadError, NotFoundError } = require('./Errors')
+const { promisify } = require('util')
+
+const pipeline = promisify(Stream.pipeline)
+
+// Observes data that passes through and computes some metadata for it
+// - specifically, it computes the number of bytes transferred, and optionally
+//   computes a cryptographic hash based on the 'hash' option. e.g., pass
+//   { hash: 'md5' } to compute the md5 hash of the stream
+// - if 'metric' is supplied as an option, this metric will be incremented by
+//   the number of bytes transferred
+class ObserverStream extends Stream.Transform {
+  constructor(options) {
+    options.autoDestroy = true
+    super(options)
```

Comment:
Just to avoid mutating the options argument:
```suggestion
    super({ autoDestroy: true, ...options })
```